File: ios/Runner/Native/Display/DisplayNativeHandler.swift
import Flutter
import UIKit

class DisplayNativeHandler: AbstractNativeHandler {
    override var channelName: String {
        return "display_service"
    }
    
    override var eventChannels: [String: String] {
        return [
            "orientation": "display_service/orientation",
            "brightness": "display_service/brightness"
        ]
    }
    
    private let provider = DisplayInfoProvider()
    private var orientationListener: OrientationListener?
    private var brightnessListener: BrightnessListener?
    
    override func handleMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getDisplayInfo":
            let info = provider.getDisplayInfo()
            result(info)
            
        case "getRefreshRate":
            let refreshRate = provider.getRefreshRate()
            result(refreshRate)
            
        case "setBrightness":
            if let brightness = call.arguments as? Double {
                let success = provider.setBrightness(brightness)
                result(success)
            } else {
                result(FlutterError(
                    code: "INVALID_ARGUMENTS",
                    message: "Brightness must be a double value between 0.0 and 1.0",
                    details: nil
                ))
            }
            
        case "isPortrait":
            result(provider.isPortrait())
            
        case "isLandscape":
            result(provider.isLandscape())
            
        case "isDarkMode":
            result(provider.isDarkMode())
            
        case "getScreenSize":
            result(provider.getScreenSize())
            
        case "getScreenInches":
            result(provider.getScreenInches())
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    override func getStreamHandler(eventName: String) -> (any FlutterStreamHandler & NSObjectProtocol)? {
        switch eventName {
        case "orientation":
            orientationListener = OrientationListener()
            return orientationListener
        case "brightness":
            brightnessListener = BrightnessListener()
            return brightnessListener
        default:
            return nil
        }
    }
    
    override func onDestroy() {
        super.onDestroy()
        orientationListener = nil
        brightnessListener = nil
    }
}

File: android/app/src/main/kotlin/com/example/news_app/native/display/DisplayNativeHandler.kt
package com.nytimes.news_app.native.display

import BrightnessListener
import OrientationListener
import android.content.Context
import android.os.Build
import androidx.annotation.RequiresApi
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.EventChannel
import com.nytimes.news_app.native.BaseNativeHandler

class DisplayNativeHandler(flutterEngine: FlutterEngine, context: Context) : BaseNativeHandler(flutterEngine, context) {

    override val channelName = "display_service"
    override val eventChannels = mapOf(
        "orientation" to "display_service/orientation",
        "brightness" to "display_service/brightness"
    )

    private val provider = DisplayInfoProvider(context)
    private var orientationListener: OrientationListener? = null
    private var brightnessListener: BrightnessListener? = null

    @RequiresApi(Build.VERSION_CODES.N)
    override fun handleMethodCall(call: MethodCall, result: Result) {
        try {
            when (call.method) {
                "getDisplayInfo" -> result.success(provider.getDisplayInfo())
                "getRefreshRate" -> result.success(provider.getRefreshRate().toDouble())
                "setBrightness" -> {
                    val brightness = call.arguments as? Double ?: 1.0
                    result.success(provider.setBrightness(brightness))
                }
                "isPortrait" -> result.success(provider.isPortrait())
                "isLandscape" -> result.success(provider.isLandscape())
                "isDarkMode" -> result.success(provider.isDarkMode())
                "getScreenSize" -> result.success(provider.getScreenSize())
                "getScreenInches" -> result.success(provider.getScreenInches())
                else -> result.notImplemented()
            }
        } catch (e: Exception) {
            result.error("NATIVE_ERROR", e.message, null)
        }
    }

    override fun getStreamHandler(eventName: String): EventChannel.StreamHandler? {
        return when (eventName) {
            "orientation" -> {
                orientationListener = OrientationListener(context)
                orientationListener
            }
            "brightness" -> {
                brightnessListener = BrightnessListener(context)
                brightnessListener
            }
            else -> null
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        orientationListener?.destroy()
        brightnessListener?.destroy()
        orientationListener = null
        brightnessListener = null
    }
}